File: src/main/java/xyz/emirdev/echogen/task/RamBarTask.java
package xyz.emirdev.echogen.task;

import lombok.Getter;
import net.kyori.adventure.bossbar.BossBar;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.minimessage.MiniMessage;
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
import org.bukkit.entity.Player;

import java.lang.management.ManagementFactory;
import java.lang.management.MemoryUsage;

public class RamBarTask extends BossBarTask {
    private static RamBarTask instance;
    @Getter
    private long allocated = 0L;
    @Getter
    private long used = 0L;
    @Getter
    private long xmx = 0L;
    @Getter
    private long xms = 0L;
    @Getter
    private float percent = 0F;
    private int tick = 0;

    public static RamBarTask instance() {
        if (instance == null) {
            instance = new RamBarTask();
        }
        return instance;
    }

    @Override
    BossBar createBossBar() {
        return BossBar.bossBar(Component.text(""), 0.0F, instance().getBossBarColor(), BossBar.Overlay.NOTCHED_20);
    }

    @Override
    void updateBossBar(BossBar bossbar, Player player) {
        bossbar.progress(getBossBarProgress());
        bossbar.color(getBossBarColor());
        bossbar.name(MiniMessage.miniMessage().deserialize("<gray>Ram<yellow>:</yellow> <used>/<xmx> (<percent>)",
                Placeholder.component("allocated", format(this.allocated)),
                Placeholder.component("used", format(this.used)),
                Placeholder.component("xmx", format(this.xmx)),
                Placeholder.component("xms", format(this.xms)),
                Placeholder.unparsed("percent", ((int) (this.percent * 100)) + "%")
        ));
    }

    @Override
    public void run() {
        if (++this.tick < 20) {
            return;
        }
        this.tick = 0;

        MemoryUsage heap = ManagementFactory.getMemoryMXBean().getHeapMemoryUsage();

        this.allocated = heap.getCommitted();
        this.used = heap.getUsed();
        this.xmx = heap.getMax();
        this.xms = heap.getInit();
        this.percent = Math.max(Math.min((float) this.used / this.xmx, 1.0F), 0.0F);

        super.run();
    }

    private float getBossBarProgress() {
        return this.percent;
    }

    private BossBar.Color getBossBarColor() {
        if (this.percent < 0.5F) {
            return BossBar.Color.GREEN;
        } else if (this.percent < 0.75F) {
            return BossBar.Color.YELLOW;
        } else {
            return BossBar.Color.RED;
        }
    }

    public Component format(long v) {
        String color;
        if (this.percent < 0.60F) {
            color = "<gradient:#55ff55:#00aa00><text></gradient>";
        } else if (this.percent < 0.85F) {
            color = "<gradient:#ffff55:#ffaa00><text></gradient>";
        } else {
            color = "<gradient:#ff5555:#aa0000><text></gradient>";
        }
        String value;
        if (v < 1024) {
            value = v + "B";
        } else {
            int z = (63 - Long.numberOfLeadingZeros(v)) / 10;
            value = String.format("%.1f%s", (double) v / (1L << (z * 10)), "BKMGTPE".charAt(z));
        }
        return MiniMessage.miniMessage().deserialize(color, Placeholder.unparsed("text", value));
    }
}

File: src/main/kotlin/cat/emir/echogen/task/RamBarTask.kt
package cat.emir.echogen.task

import cat.emir.echogen.Echogen
import cat.emir.echogen.toComponent
import net.kyori.adventure.bossbar.BossBar
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder
import org.bukkit.entity.Player

import java.lang.management.ManagementFactory
import kotlin.math.max
import kotlin.math.min

class RamBarTask private constructor(plugin: Echogen) : BossBarTask(plugin) {
    companion object {
        private var instance: RamBarTask? = null

        fun instance(plugin: Echogen): RamBarTask {
            return instance ?: RamBarTask(plugin).also { instance = it }
        }
    }
    var allocated = 0L
    var used = 0L
    var xmx = 0L
    var xms = 0L
    var percent = 0F
    var tick = 0

    override fun createBossBar(): BossBar {
        return BossBar.bossBar(Component.text(""), 0.0F, instance(plugin).getBossBarColor(), BossBar.Overlay.NOTCHED_20)
    }

    override fun updateBossBar(bossbar: BossBar, player: Player) {
        bossbar.progress(getBossBarProgress())
        bossbar.color(getBossBarColor())
        bossbar.name("<gray>Ram<yellow>:</yellow> <used>/<xmx> (<percent>)".toComponent(
            Placeholder.component("allocated", format(this.allocated)),
            Placeholder.component("used", format(this.used)),
            Placeholder.component("xmx", format(this.xmx)),
            Placeholder.component("xms", format(this.xms)),
            Placeholder.unparsed("percent", (this.percent * 100).toInt().toString() + "%")
        ))
    }

    override fun run() {
        if (++this.tick < 20) {
            return
        }
        this.tick = 0

        val heap = ManagementFactory.getMemoryMXBean().heapMemoryUsage

        this.allocated = heap.committed
        this.used = heap.used
        this.xmx = heap.max
        this.xms = heap.init
        this.percent = (this.used.toFloat() / this.xmx).coerceIn(0.0F, 1.0F)

        super.run()
    }

    fun getBossBarProgress(): Float {
        return this.percent
    }

    fun getBossBarColor(): BossBar.Color {
        return if (this.percent < 0.5F) BossBar.Color.GREEN
        else if (this.percent < 0.75F) BossBar.Color.YELLOW
        else BossBar.Color.RED
    }

    fun format(v: Long): Component {
        val color = if (this.percent < 0.60F) "<gradient:#55ff55:#00aa00><text></gradient>"
        else if (this.percent < 0.85F) "<gradient:#ffff55:#ffaa00><text></gradient>"
        else "<gradient:#ff5555:#aa0000><text></gradient>"

        val value = if (v < 1024) {
            "${v}B"
        } else {
            val z = (63 - v.countLeadingZeroBits()) / 10
            "%.1f%s".format(v.toDouble() / (1L shl (z * 10)), "BKMGTPE"[z])
        }

        return color.toComponent(Placeholder.unparsed("text", value))
    }
}